### PR TITLE
CR-1101262 add clock scaling feature configs in xbmgmt dump

### PR DESCRIPTION
**Pull request by rajkumar-xilinx**
Signed-off-by: Rajkumar Rampelli <rajkumar@xilinx.com>

### CODE REVIEW

**Comment by rajkumar-xilinx** [comment]
Output looks like below on u2 setup.

$xbmgmt dump --device 0000:85:00.0 -c -o out2.ini
config has been dumped to out2.ini

$cat out2.ini
[Device]
mailbox_channel_disable=0x0
mailbox_channel_switch=0x0
cache_xclbin=0
security_level=0
scaling_support=true
scaling_enabled=true
scaling_critical_power_threshold=N/A
scaling_critical_temp_threshold=97C
scaling_threshold_power_limit=25W
scaling_threshold_temp_limit=93C
scaling_threshold_power_override_en=1
scaling_threshold_power_override=1W
scaling_threshold_temp_override_en=1
scaling_threshold_temp_override=55C
data_retention=4294967295

**On U30 setup:**
$xbmgmt dump --device 0000:5f:00.0 -c -o out3.ini
config has been dumped to out3.ini

$cat out3.ini
[Device]
mailbox_channel_disable=0x0
mailbox_channel_switch=0x0
cache_xclbin=0
security_level=0

**Comment by rajkumar-xilinx** [comment]
retest this please

File: src/runtime_src/core/common/query_requests.h [hunk 1]
```diff
@@ -96,9 +96,18 @@ enum class key_type
...
+  xmc_scaling_support,
   xmc_scaling_enabled,
   xmc_scaling_override,
   xmc_scaling_temp_override,
+  xmc_scaling_critical_pow_threshold,
+  xmc_scaling_critical_temp_threshold,
+  xmc_scaling_threshold_power_limit,
+  xmc_scaling_threshold_temp_limit,
+  xmc_scaling_threshold_power_override_en,
+  xmc_scaling_threshold_temp_override_en,
```
Comment (AShivangi) [inline src/runtime_src/core/common/query_requests.h:108]: what does "en" stand for?

File: src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp [hunk 2]
```diff
@@ -52,6 +52,29 @@ flash_dump(const std::shared_ptr<xrt_core::device>& _dev, const std::string outp
   flasher.readBack(output);
 }
 
+static bool
+isSupported(const std::shared_ptr<xrt_core::device>& dev)
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp:56]: no camel case please `is_supported()`

File: src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp [hunk 3]
```diff
@@ -52,6 +52,29 @@ flash_dump(const std::shared_ptr<xrt_core::device>& _dev, const std::string outp
   flasher.readBack(output);
 }
 
+static bool
+isSupported(const std::shared_ptr<xrt_core::device>& dev)
+{
+    std::string errmsg;
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp:58]: I don't think this is used anywhere

File: src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp [hunk 4]
```diff
@@ -52,6 +52,29 @@ flash_dump(const std::shared_ptr<xrt_core::device>& _dev, const std::string outp
   flasher.readBack(output);
 }
 
+static bool
+isSupported(const std::shared_ptr<xrt_core::device>& dev)
+{
+    std::string errmsg;
+    bool is_mfg = false, is_recovery = false;
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp:59]: Separate lines

File: src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp [hunk 5]
```diff
@@ -52,6 +52,29 @@ flash_dump(const std::shared_ptr<xrt_core::device>& _dev, const std::string outp
...
 
+static bool
+isSupported(const std::shared_ptr<xrt_core::device>& dev)
+{
+    std::string errmsg;
+    bool is_mfg = false, is_recovery = false;
+
+    try {
+      is_mfg = xrt_core::device_query<xrt_core::query::is_mfg>(dev);
+    } catch(...) {}
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp:63]: Please don't catch (...), for query errors, use:
`const xrt_core::query::exception&`

File: src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp [hunk 6]
```diff
@@ -52,6 +52,29 @@ flash_dump(const std::shared_ptr<xrt_core::device>& _dev, const std::string outp
...
+
+    try {
+      is_recovery = xrt_core::device_query<xrt_core::query::is_recovery>(dev);
+    }
+    catch(...) {}
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp:68]: ditto

File: src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp [hunk 7]
```diff
@@ -52,6 +52,29 @@ flash_dump(const std::shared_ptr<xrt_core::device>& _dev, const std::string outp
   flasher.readBack(output);
 }
 
+static bool
+isSupported(const std::shared_ptr<xrt_core::device>& dev)
+{
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp:57]: The indentation of this function looks off

File: src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp [hunk 8]
```diff
@@ -68,14 +91,32 @@ config_dump(const std::shared_ptr<xrt_core::device>& _dev, const std::string out
   child.put("mailbox_channel_disable", xrt_core::device_query<xrt_core::query::config_mailbox_channel_disable>(_dev));
   child.put("mailbox_channel_switch", xrt_core::device_query<xrt_core::query::config_mailbox_channel_switch>(_dev));
   child.put("cache_xclbin", xrt_core::device_query<xrt_core::query::cache_xclbin>(_dev));
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp:93]: Are these supported in mfg mode?

File: src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp [hunk 9]
```diff
@@ -68,14 +91,32 @@ config_dump(const std::shared_ptr<xrt_core::device>& _dev, const std::string out
...
+      child.put("scaling_critical_temp_threshold", xrt_core::device_query<xrt_core::query::xmc_scaling_critical_temp_threshold>(_dev));
+      child.put("scaling_threshold_power_limit", xrt_core::device_query<xrt_core::query::xmc_scaling_threshold_power_limit>(_dev));
+      child.put("scaling_threshold_temp_limit", xrt_core::device_query<xrt_core::query::xmc_scaling_threshold_temp_limit>(_dev));
+      child.put("scaling_threshold_power_override_en", xrt_core::device_query<xrt_core::query::xmc_scaling_threshold_power_override_en>(_dev));
+      child.put("scaling_threshold_power_override", xrt_core::device_query<xrt_core::query::xmc_scaling_threshold_power_override>(_dev));
+      child.put("scaling_threshold_temp_override_en", xrt_core::device_query<xrt_core::query::xmc_scaling_threshold_temp_override_en>(_dev));
+      child.put("scaling_threshold_temp_override", xrt_core::device_query<xrt_core::query::xmc_scaling_threshold_temp_override>(_dev));
+      child.put("data_retention", xrt_core::device_query<xrt_core::query::data_retention>(_dev));
+    } catch(...) {}
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp:109]: Ditto

**Review by AShivangi — CHANGES_REQUESTED** [review]
Hi @rajkumar-xilinx, 
Why does the output for u.2 and u30 differ?

File: src/runtime_src/core/common/query_requests.h (hunk 1 above)
Comment (rajkumar-xilinx) [reply src/runtime_src/core/common/query_requests.h:108]: enable

**Comment by rajkumar-xilinx** [comment]
> Hi @rajkumar-xilinx,
> Why does the output for u.2 and u30 differ?

clock scaling supported on U2, so this feature configs are not available in u30

File: src/runtime_src/core/common/query_requests.h (hunk 1 above)
Comment (AShivangi) [reply src/runtime_src/core/common/query_requests.h:109]: So this value would be true or false?
I understand that the name is already very long, but it is ok to be verbose here. Please change en to enable

File: src/runtime_src/core/common/query_requests.h [hunk 10]
```diff
@@ -1003,6 +1012,87 @@ struct xmc_reg_base : request
...
+struct xmc_scaling_threshold_temp_limit : request
+{
+  using result_type = std::string;       // get value type
+  static const key_type key = key_type::xmc_scaling_threshold_temp_limit;
+
+  virtual boost::any
+  get(const device*) const = 0;
+};
+
+struct xmc_scaling_threshold_power_override_en : request
```
Comment (AShivangi) [inline src/runtime_src/core/common/query_requests.h:1060]: If enable is true/false, please change the return type to bool. Same for the following function. We have some examples in this file for reference

File: src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp [hunk 11]
```diff
@@ -52,6 +52,28 @@ flash_dump(const std::shared_ptr<xrt_core::device>& _dev, const std::string outp
...
+  try {
+    is_mfg = xrt_core::device_query<xrt_core::query::is_mfg>(dev);
+  } catch(const xrt_core::query::exception&) {}
+
+  try {
+    is_recovery = xrt_core::device_query<xrt_core::query::is_recovery>(dev);
+  } catch(const xrt_core::query::exception&) {}
+
+  if (is_mfg || is_recovery) {
+    std::cerr << "This operation is not supported with manufacturing image. mfg:recovery=" << is_mfg << ":" << is_recovery << std::endl;
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdDump.cpp:70]:
```suggestion
    std::cerr << boost::format("This operation is not supported with %s image.\n) % (is_mfg ? "manufacturing" : "recovery");

```

**Review by AShivangi — CHANGES_REQUESTED** [review]
Hi @rajkumar-xilinx, 
Why does the output for u.2 and u30 differ?

**Comment by AShivangi** [comment]
Thanks, @chvamshi-xilinx for pointing this out. I wasn't sure where the non-user changeable values should go. "-r platform" sounds the right place for them. 

I think "data_retention" might be the ddr retention, but the value looks off. 

FYI, I am working on implementing **xbmgmt configure --load-config** so if you make changes to **OO_LoadConfig**, our changes might overlap so keep an eye out

**Comment by chvamshi-xilinx** [comment]
> Thanks, @chvamshi-xilinx for pointing this out. I wasn't sure where the non-user changeable values should go. "-r platform" sounds the right place for them.
> 
> I think "data_retention" might be the ddr retention, but the value looks off.
> 
> FYI, I am working on implementing **xbmgmt configure --load-config** so if you make changes to **OO_LoadConfig**, our changes might overlap so keep an eye out

Thanks @AShivangi . @rajkumar-xilinx  is adding changes to "xbutil examine -r CMC"  as clock scaling is controlled by CMC. 
Other user changeable options will be part of dump. He is going to add support them to load back using --load-config also.

**Comment by rajkumar-xilinx** [comment]
Following the sequence which demonstrates the reading and configuring the CMC clock scaling feature using xbmgmt2 tool.

```
$xbmgmt examine -r cmc --device 0000:85:00.0
-----------------------------------------------------
1/1 [0000:85:00.0] : xilinx_u2_gen3x4_xdma_gc_base_2
-----------------------------------------------------
CMC
  Runtime clock scaling feature :
        Supported              : true
        Enabled                : true
        Critical threshold (clock shutdown) limits:
                Power                  : N/A
                Temperature            : 97
        Throttling threshold limits:
                Power                  : 25
                Temperature            : 93
        Power threshold override:
                Override               : true
                Override limit         : 25
        Temperature threshold override:
                Override               : true
                Override limit         : 93

$xbmgmt dump --device 0000:85:00.0 -c -o out7.ini
config has been dumped to out7.ini
$cat out7.ini
[Device]
mailbox_channel_disable=0x0
mailbox_channel_switch=0x0
cache_xclbin=0
scaling_enabled=true
scaling_power_override=25
scaling_temp_override=93

$vim out7.ini  --> edited to set power to 20 and temp to 80

$xbmgmt advanced --load-config out7.ini --device 0000:85:00.0
config has been successfully loaded

$xbmgmt examine -r cmc --device 0000:85:00.0
-----------------------------------------------------
1/1 [0000:85:00.0] : xilinx_u2_gen3x4_xdma_gc_base_2
-----------------------------------------------------
CMC
  Runtime clock scaling feature :
        Supported              : true
        Enabled                : false
        Critical threshold (clock shutdown) limits:
                Power                  : N/A
                Temperature            : 97
        Throttling threshold limits:
                Power                  : 25
                Temperature            : 93
        Power threshold override:
                Override               : true
                Override limit         : 20
        Temperature threshold override:
                Override               : true
                Override limit         : 80

```

File: src/runtime_src/core/tools/xbmgmt2/SubCmdExamine.cpp [hunk 12]
```diff
@@ -51,7 +52,7 @@ static const ReportCollection fullReportCollection = {
   #ifdef ENABLE_NATIVE_SUBCMDS_AND_REPORTS
     std::make_shared<ReportMechanical>(),
     std::make_shared<ReportFirewall>(),
-    std::make_shared<ReportMailbox>()
+    std::make_shared<ReportCmcStatus>()
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdExamine.cpp:55]: ReportMailbox is accidently removed?

File: src/runtime_src/core/tools/common/ReportCmcStatus.cpp [hunk 13]
```diff
@@ -34,14 +34,26 @@ ReportCmcStatus::getPropertyTree20202( const xrt_core::device * _pDevice,
...
+  pt.put("Description", "CMC");
   try {
-    pt.put("Description", "CMC status");
     pt.put("heartbeat_err_time", xrt_core::device_query<xrt_core::query::heartbeat_err_time>(_pDevice));
     pt.put("heartbeat_count", xrt_core::device_query<xrt_core::query::heartbeat_count>(_pDevice));
     pt.put("heartbeat_err_code", xrt_core::device_query<xrt_core::query::heartbeat_err_code>(_pDevice));
     pt.put("heartbeat_stall", xrt_core::device_query<xrt_core::query::heartbeat_stall>(_pDevice));
     pt.put("status", xrt_core::utils::parse_cmc_status(static_cast<unsigned int>(xrt_core::device_query<xrt_core::query::heartbeat_err_code>(_pDevice))));
-  } catch(const xrt_core::query::no_such_key&) {}
+  } catch(...) {}
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/common/ReportCmcStatus.cpp:44]: Why do we need to catch all exceptions? We may just need to catch "no_such_key" ? Any issues with it?

File: src/runtime_src/core/tools/common/ReportCmcStatus.cpp [hunk 14]
```diff
@@ -34,14 +34,26 @@ ReportCmcStatus::getPropertyTree20202( const xrt_core::device * _pDevice,
...
+    pt.put("scaling_support", xrt_core::device_query<xrt_core::query::xmc_scaling_support>(_pDevice));
+    pt.put("scaling_critical_power_threshold", xrt_core::device_query<xrt_core::query::xmc_scaling_critical_pow_threshold>(_pDevice));
+    pt.put("scaling_critical_temp_threshold", xrt_core::device_query<xrt_core::query::xmc_scaling_critical_temp_threshold>(_pDevice));
+    pt.put("scaling_threshold_power_limit", xrt_core::device_query<xrt_core::query::xmc_scaling_threshold_power_limit>(_pDevice));
+    pt.put("scaling_threshold_temp_limit", xrt_core::device_query<xrt_core::query::xmc_scaling_threshold_temp_limit>(_pDevice));
+    pt.put("scaling_power_override_en", xrt_core::device_query<xrt_core::query::xmc_scaling_power_override_enable>(_pDevice));
+    pt.put("scaling_power_override", xrt_core::device_query<xrt_core::query::xmc_scaling_power_override>(_pDevice));
+    pt.put("scaling_temp_override_en", xrt_core::device_query<xrt_core::query::xmc_scaling_temp_override_enable>(_pDevice));
+    pt.put("scaling_temp_override", xrt_core::device_query<xrt_core::query::xmc_scaling_temp_override>(_pDevice));
+  } catch(...) {}
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/common/ReportCmcStatus.cpp:56]: Same about "catch" here also..

**Review by chvamshi-xilinx — CHANGES_REQUESTED** [review]
Overall it looks good now. Just couple of minor comments.

**Comment by xilinxgitops** [comment]
retest this please

File: src/runtime_src/core/tools/common/ReportCmcStatus.cpp [hunk 15]
```diff
@@ -34,14 +34,30 @@ ReportCmcStatus::getPropertyTree20202( const xrt_core::device * _pDevice,
...
-    pt.put("Description", "CMC status");
     pt.put("heartbeat_err_time", xrt_core::device_query<xrt_core::query::heartbeat_err_time>(_pDevice));
     pt.put("heartbeat_count", xrt_core::device_query<xrt_core::query::heartbeat_count>(_pDevice));
     pt.put("heartbeat_err_code", xrt_core::device_query<xrt_core::query::heartbeat_err_code>(_pDevice));
     pt.put("heartbeat_stall", xrt_core::device_query<xrt_core::query::heartbeat_stall>(_pDevice));
     pt.put("status", xrt_core::utils::parse_cmc_status(static_cast<unsigned int>(xrt_core::device_query<xrt_core::query::heartbeat_err_code>(_pDevice))));
-  } catch(const xrt_core::query::no_such_key&) {}
+  }
+  catch(const xrt_core::query::no_such_key&) {}
+  catch(const xrt_core::query::sysfs_error&) {}
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/ReportCmcStatus.cpp:46]:
```suggestion
  catch(const xrt_core::query::exception&) {}
```
Soren introduced this new exception which catches numerous key related issues

File: src/runtime_src/core/tools/common/ReportCmcStatus.cpp [hunk 16]
```diff
@@ -34,14 +34,30 @@ ReportCmcStatus::getPropertyTree20202( const xrt_core::device * _pDevice,
...
+    pt.put("scaling_critical_temp_threshold", xrt_core::device_query<xrt_core::query::xmc_scaling_critical_temp_threshold>(_pDevice));
+    pt.put("scaling_threshold_power_limit", xrt_core::device_query<xrt_core::query::xmc_scaling_threshold_power_limit>(_pDevice));
+    pt.put("scaling_threshold_temp_limit", xrt_core::device_query<xrt_core::query::xmc_scaling_threshold_temp_limit>(_pDevice));
+    pt.put("scaling_power_override_en", xrt_core::device_query<xrt_core::query::xmc_scaling_power_override_enable>(_pDevice));
+    pt.put("scaling_power_override", xrt_core::device_query<xrt_core::query::xmc_scaling_power_override>(_pDevice));
+    pt.put("scaling_temp_override_en", xrt_core::device_query<xrt_core::query::xmc_scaling_temp_override_enable>(_pDevice));
+    pt.put("scaling_temp_override", xrt_core::device_query<xrt_core::query::xmc_scaling_temp_override>(_pDevice));
+  }
+  catch(const xrt_core::query::no_such_key&) {}
+  catch(const xrt_core::query::sysfs_error&) {}
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/ReportCmcStatus.cpp:60]: ditto

File: src/runtime_src/core/tools/common/ReportCmcStatus.cpp [hunk 17]
```diff
@@ -34,14 +34,30 @@ ReportCmcStatus::getPropertyTree20202( const xrt_core::device * _pDevice,
...
+    pt.put("scaling_support", xrt_core::device_query<xrt_core::query::xmc_scaling_support>(_pDevice));
+    pt.put("scaling_critical_power_threshold", xrt_core::device_query<xrt_core::query::xmc_scaling_critical_pow_threshold>(_pDevice));
+    pt.put("scaling_critical_temp_threshold", xrt_core::device_query<xrt_core::query::xmc_scaling_critical_temp_threshold>(_pDevice));
+    pt.put("scaling_threshold_power_limit", xrt_core::device_query<xrt_core::query::xmc_scaling_threshold_power_limit>(_pDevice));
+    pt.put("scaling_threshold_temp_limit", xrt_core::device_query<xrt_core::query::xmc_scaling_threshold_temp_limit>(_pDevice));
+    pt.put("scaling_power_override_en", xrt_core::device_query<xrt_core::query::xmc_scaling_power_override_enable>(_pDevice));
+    pt.put("scaling_power_override", xrt_core::device_query<xrt_core::query::xmc_scaling_power_override>(_pDevice));
+    pt.put("scaling_temp_override_en", xrt_core::device_query<xrt_core::query::xmc_scaling_temp_override_enable>(_pDevice));
+    pt.put("scaling_temp_override", xrt_core::device_query<xrt_core::query::xmc_scaling_temp_override>(_pDevice));
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/ReportCmcStatus.cpp:57]: This should be more layered:
```
"scaling" : {
     "enabled": "true",
     ...
    "critical_threshold_limits: {
        power_watts: "N/A",
        temp_C = "97"
    }
    ...
}
```

Group them like how you have grouped them while printing. When you make these changes, please add the json output to the PR for review. 
Please add "_watts" or "_C" as well to the node names

File: src/runtime_src/core/tools/common/ReportCmcStatus.cpp [hunk 18]
```diff
@@ -54,14 +70,33 @@ ReportCmcStatus::writeReport( const xrt_core::device* /*_pDevice*/,
...
+  } catch(...) {}
+  try {
+    _output << boost::format("  %-22s :\n") % "Runtime clock scaling feature";
+    _output << boost::format("  \t%-22s : %s\n") % "Supported" % cmc.get<std::string>("scaling_support");
+    _output << boost::format("  \t%-22s : %s\n") % "Enabled" % cmc.get<std::string>("scaling_enabled");
+    _output << boost::format("  \t%-22s:\n") % "Critical threshold (clock shutdown) limits";
+    _output << boost::format("  \t\t%-22s : %s\n") % "Power" % cmc.get<std::string>("scaling_critical_power_threshold");
+    _output << boost::format("  \t\t%-22s : %s\n") % "Temperature" % cmc.get<std::string>("scaling_critical_temp_threshold");
+    _output << boost::format("  \t%-22s:\n") % "Throttling threshold limits";
+    _output << boost::format("  \t\t%-22s : %s\n") % "Power" % cmc.get<std::string>("scaling_threshold_power_limit");
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/ReportCmcStatus.cpp:93]:
```suggestion
    _output << boost::format("  \t\t%-22s : %s Watts\n") % "Power" % cmc.get<std::string>("scaling_threshold_power_limit");

```

Please add watts and C to the outputs

File: src/runtime_src/core/tools/common/ReportCmcStatus.cpp [hunk 19]
```diff
@@ -54,14 +70,33 @@ ReportCmcStatus::writeReport( const xrt_core::device* /*_pDevice*/,
...
-    _output << boost::format("  %-22s : %s sec\n\n") % "err time" % cmc.get<std::string>("heartbeat_err_time");
+  try {
+    uint32_t err_code = cmc.get<uint32_t>("heartbeat_err_code");
+    _output << boost::format("  %-22s : 0x%x %s\n") % "Status" % err_code % cmc.get<std::string>("status");
+    if (err_code)
+      _output << boost::format("  %-22s : %s sec\n\n") % "err time" % cmc.get<std::string>("heartbeat_err_time");
+  } catch(...) {}
+  try {
+    _output << boost::format("  %-22s :\n") % "Runtime clock scaling feature";
+    _output << boost::format("  \t%-22s : %s\n") % "Supported" % cmc.get<std::string>("scaling_support");
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/ReportCmcStatus.cpp:87]:
```suggestion
    _output << boost::format("    %-22s : %s\n") % "Supported" % cmc.get<std::string>("scaling_support");
```
Please don't use tab. 2 spaces is good enough :)

**Review by AShivangi — APPROVED** [review]
A few additional comments, but this is looking pretty good now!

File: src/runtime_src/core/tools/common/ReportCmcStatus.cpp (hunk 15 above)
Comment (rajkumar-xilinx) [reply src/runtime_src/core/tools/common/ReportCmcStatus.cpp:46]: Hi @AShivangi  the error is related to sysfs node absence. So, this exception will not catch this error. sysfs_error is the right catch.

**Comment by rajkumar-xilinx** [comment]
**$ xbutil examine -r cmc -f JSON -o out7.json -d 0000:d8:00.1**
```
------------------------------------------------
1/1 [0000:d8:00.1] : xilinx_u250_gen3x16_base_3
------------------------------------------------
CMC
  Status                 : 0x0 (GOOD)
  Runtime clock scaling feature :
    Supported : false
    Enabled : false
    Critical threshold (clock shutdown) limits:
      Power : 0 W
      Temperature : 0 C
    Throttling threshold limits:
      Power : 0 W
      Temperature : 0 C
    Power threshold override:
      Override : false
      Override limit : 0 W
    Temperature threshold override:
      Override : false
      Override limit : 0 C
Successfully wrote the JSON file: out7.json

```
**$ cat out7.json**

```
{
    "schema_version":
    {
        "schema": "JSON",
        "creation_date": "Mon Sep  6 20:15:49 2021 GMT"
    },
    "devices":
    [
        {
            "interface_type": "pcie",
            "device_id": "0000:d8:00.1",
            "cmc":
            {
                "Description": "CMC",
                "cmc_heartbeat":
                {
                    "Description": "CMC heartbeat",
                    "heartbeat_err_time": "0",
                    "heartbeat_count": "265",
                    "heartbeat_err_code": "0",
                    "heartbeat_stall": "0",
                    "status": "(GOOD)"
                },
                "scaling":
                {
                    "Description": "Runtime Clock Scaling",
                    "enabled": "false",
                    "supported": "false",
                    "shutdown_threshold_limits":
                    {
                        "power_watts": "0",
                        "temp_celsius": "0"
                    },
                    "override_threshold_limits":
                    {
                        "power_watts": "0",
                        "temp_celsius": "0"
                    },
                    "power_threshold_override":
                    {
                        "enabled": "false",
                        "power_watts": "0"
                    },
                    "temp_threshold_override":
                    {
                        "enabled": "false",
                        "temp_celsius": "0"
                    }
                }
            }
        }
    ]
}

```

**Review by chvamshi-xilinx — APPROVED** [review]
Looks good.

**Review by stsoe — APPROVED** [review]

**Pull request merged**